tests/e2e/utils/user.js

```diff
@@ -0,0 +1,45 @@
+import { expect } from '@playwright/test';
+
+/**
+ * Logs in a user with the given credentials on the provided page, with retries if login fails.
+ * @param {Object} page - The Playwright page object to use for the login process.
+ * @param {string} username - The username to use for the login process.
+ * @param {string} password - The password to use for the login process.
+ * @param {number} [retries=3] - The number of retries for the login process in case of failure.
+ * @throws {Error} Will throw an error if login fails after the specified number of retries.
+ * @returns {Promise<void>} - A promise that resolves when the login process is complete.
+ */
+export async function login( page, username, password, retries = 3 ) {
+	for ( let i = 1; i <= retries; i++ ) {
+		try {
+			await page.goto( `/wp-admin` );
+			await page.waitForLoadState( 'networkidle' );
+
+			if ( await page.url().includes( 'wp-login.php' ) ) {
+				await page.fill( 'input[name="log"]', username );
+				await page.fill( 'input[name="pwd"]', password );
+				await page.click( 'text=Log In' );
+			}
+			await page.waitForLoadState( 'networkidle' );
+
+			if ( await page.$( 'body.logged-in' ) ) {
+				// customer login
+				return;
+			} else {
+				// admin login
+				await expect( page.locator( 'div.wrap > h1' ) ).toHaveText(
+					'Dashboard'
+				);
+				return;
+			}
```

Comment:
Do you think we need a way to specifically assert whether we did a customer login or admin login? For example, we use the `user.login()` function wanting to log in a customer but it actually logs in an admin; is that something we'd want to catch and have the test fail, or does it not matter?

I think it doesn't matter since admins can also act as customers, so really just wondering if you think it'd be useful in this helper?

Reply:
> Do you think we need a way to specifically assert whether we did a customer login or admin login?

No, I don't think we need this at the moment. The users on tests are created via API, so we know their site roles for sure. And even if that wasn't the case, as you said, admins can act as customers, thus no harm to the tests. 

A more problematic mistake would be adding credentials for a customer in the env file instead of credentials for an admin. However, the execution would stop in the setup since we wouldn't be able to generate the API keys.

Reply:
Awesome, then we just keep this as is 🙂